### PR TITLE
Toggle Group Control: add tooltip

### PR DESCRIPTION
**Pull request by ramonjd**
Related:

- https://github.com/WordPress/gutenberg/issues/36545

## Description

Hey ho.

This PR takes a stab at adding tooltips to the font size segmented control.

The gist is that we're giving the user additional information about their font-size selection only otherwise available via `aria-label` (and not immediately visible in the browser).

Here's a sneak preview:

Screen Shot 2021-11-22 at 9 05 41 pm

Screen Shot 2021-11-22 at 9 05 20 pm

Why? At present the font size segmented control displays numerical values that are not the same as their descriptions. 

So `42` is `'Huge'`. 

```js
		{
			name: _x( 'Huge', 'font size name' ),
			size: 42,
			slug: 'huge',
		},
```

To achieve what we need, we've extended `ToggleGroupControlOption` to add a Tooltip wrapper around the radio component in the presence of a `tooltipText` prop.

There might be other use cases for this, for example, adding extra descriptive power 🚀 or visible hints to options. The limit is your imagination!

## How has this been tested?
`npm run storybook:dev`

Take a look at `FontSizePicker` and `ToggleGroupControl` (With Tooltip variation) : check that the tooltip appears.

Check that there are no regressions for existing usage of `ToggleGroupControlOption`. For example, inset a Navigation Block into the editor. The Overlay menu in the right-hand sidebar should look and work as expected. There should be no tooltips here.

Screen Shot 2021-11-22 at 9 56 52 pm

Another one is the Post Featured Image Block. Adjust the height dimensions and check the options group:

Screen Shot 2021-11-23 at 9 19 30 am

Run the unit test:
`npm run test-unit packages/components/src/toggle-group-control/test/index.js`

## Types of changes
Adding a feature to an existing component.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/components/src/toggle-group-control/toggle-group-control-option/component.tsx [hunk 1]
```diff
@@ -20,13 +20,28 @@ import {
...
-import type { ToggleGroupControlOptionProps } from '../types';
+import type { ToggleGroupControlOptionProps, WithToolTipProps } from '../types';
 import { useToggleGroupControlContext } from '../context';
 import * as styles from './styles';
 import { useCx } from '../../utils/hooks';
+import Tooltip from '../../tooltip';
 
 const { ButtonContentView, LabelPlaceholderView, LabelView } = styles;
 
+const WithToolTip = ( {
```
Comment (ramonjd) [inline packages/components/src/toggle-group-control/toggle-group-control-option/component.tsx]: Would this benefit from a new snapshot test? 

I tried it out and it renders the Popover frame only, nothing really related to the tooltip value.

**Comment by mtias** [comment]
Cool, thanks for looking into this!

File: packages/components/src/font-size-picker/index.js [hunk 2]
```diff
@@ -186,6 +186,7 @@ function FontSizePicker(
 								value={ option.value }
 								label={ option.label }
 								aria-label={ option.name }
+								tooltipText={ option.name }
```
Comment (mtias) [inline packages/components/src/font-size-picker/index.js]: Should this be just a boolean, then reading from aria-label?

Comment (ramonjd) [reply packages/components/src/font-size-picker/index.js]: I did have that in my first iteration, but changed it because a little voice inside my head asked: "_What if the aria label for the radio input button and tooltip text need to be different, for example, because we need to tell assisted technologies something else?"_

Also `aria-label` is optional on `ToggleGroupControlOption` so explicitly setting the `tooltipText` to determine both the display and content states at once seemed legit at the time.

To be honest, my reasoning wasn't more profound than that. 

Or is starting with Boolean and using `props[ 'aria-label' ] ?? label` for the text, and then and worrying about the other stuff later the safer approach?

Comment (ciampo) [reply packages/components/src/font-size-picker/index.js]: I also think that having a `showTooltip` boolean prop instead would be better, and I think that the proposed `props[ 'aria-label' ] ?? label` value for the tooltip text is definitely a good start

File: packages/components/src/toggle-group-control/toggle-group-control-option/component.tsx (hunk 1 above)
Comment (ciampo) [reply packages/components/src/toggle-group-control/toggle-group-control-option/component.tsx]: A better alternative could be a unit test where we try to toggle the Popover and check for it to be effectively rendered

File: packages/components/src/toggle-group-control/toggle-group-control-option/component.tsx [hunk 3]
```diff
@@ -20,13 +20,28 @@ import {
...
+	text,
+	children,
+}: WordPressComponentProps< WithToolTipProps, 'div' > ) => {
```
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control-option/component.tsx]: I don't believe there's a need to use `WordPressComponentProps` here

```suggestion
}: WithToolTipProps ) => {
```

File: packages/components/src/tooltip/index.js [hunk 4]
```diff
@@ -96,7 +96,7 @@ function Tooltip( {
 	children,
 	position,
 	text,
-	shortcut,
+	shortcut = null,
```
Comment (ciampo) [inline packages/components/src/tooltip/index.js]: Why was this change necessary?

File: packages/components/src/font-size-picker/index.js (hunk 2 above)
Comment (ntsekouras) [reply packages/components/src/font-size-picker/index.js]: I'd echo Matías about the `boolean` and use `radioProps[ 'aria-label' ] ?? label`.

**Review by ciampo — COMMENTED** [review]
Testing with Storybook, I noticed that the tooltip appears only when using the keyboard — is this by design?

https://user-images.githubusercontent.com/1083581/143001518-a2919c51-b2c2-4009-b6c8-01f717c355b0.mp4

I'm not sure if wrapping a `ToggleGroupControlOption` in a `Tooltip` affects the semantics / accessibility of the component. From a quick look everything seems ok, but it would be great if @diegohaz would be able to offer any further insights.

Also, these changes on `ToggleGroupControl` should be reflected in the README and in the `@wordpress/components` package CHANGELOG.

Finally, some time ago we had introduced a new `Flyout` component which was supposed to replace `Popover` in the long run, but I'm not sure if we should be using it. Probably another point where @diegohaz can give some advice.

File: packages/components/src/tooltip/index.js (hunk 4 above)
Comment (ramonjd) [reply packages/components/src/tooltip/index.js]: The TS was complaining that `shortcut` was a required prop, but, looking at all usages of Tooltip, it isn't.

So until we type Tooltip in another PR, I thought this was the least offensive way to make it happy.

File: packages/components/src/font-size-picker/index.js (hunk 2 above)
Comment (ramonjd) [reply packages/components/src/font-size-picker/index.js]: Thanks for the confidence check, folks. I'll implement as suggested.

**Comment by ramonjd** [comment]
Thanks for testing, everyone! 🙇‍♂️ 

> Testing with Storybook, I noticed that the tooltip appears only when using the keyboard — is this by design?

The tooltip [appears onMouseover, and with a bit of delay.](https://github.com/WordPress/gutenberg/blob/1b5367780f3446441bc33daf4d2c59349ce87b65/packages/components/src/tooltip/README.md)

I double-checked in FF, Safari and Chrome and it behaves as expected for me. Let me know if the problem persists on your side though and I can test further.

Kapture 2021-11-24 at 08 18 24

> Also, these changes on ToggleGroupControl should be reflected in the README and in the @wordpress/components package CHANGELOG.

Ah got it! Thanks for the reminder!

**Comment by ramonjd** [comment]
I've updated the prop, tests, readme and changelog in 761dacf5a0aa8598bc3138d71dd3de3b5e299025

Thanks for the feedback again!

File: packages/components/src/toggle-group-control/stories/index.js [hunk 5]
```diff
@@ -71,6 +71,11 @@ const _default = ( { options } ) => {
 				opt[ 'aria-label' ],
 				`${ KNOBS_GROUPS.ToggleGroupControlOption }-${ index + 1 }`
 			) }
+			tooltipText={ text(
+				`${ KNOBS_GROUPS.ToggleGroupControlOption }: tooltipText`,
+				opt.tooltipText,
+				`${ KNOBS_GROUPS.ToggleGroupControlOption }-${ index + 1 }`
+			) }
```
Comment (aaronrobertshaw) [inline packages/components/src/toggle-group-control/stories/index.js]: I think this was missed in https://github.com/WordPress/gutenberg/pull/36726/commits/761dacf5a0aa8598bc3138d71dd3de3b5e299025 when switching to use `showTooltip` instead.

```suggestion
			showTooltip={ boolean(
				`${ KNOBS_GROUPS.ToggleGroupControlOption }: showTooltip`,
				opt.showTooltip,
				`${ KNOBS_GROUPS.ToggleGroupControlOption }-${ index + 1 }`
			) }
```

File: packages/components/src/toggle-group-control/toggle-group-control-option/component.tsx [hunk 6]
```diff
@@ -54,23 +77,26 @@ function ToggleGroupControlOption(
 		className,
 		isActive && styles.buttonActive
 	);
+	const optionLabel = radioProps[ 'aria-label' ] ?? label;
```
Comment (aaronrobertshaw) [inline packages/components/src/toggle-group-control/toggle-group-control-option/component.tsx]: The null coalescing operator here will allow an empty string in the `aria-label` prop to be used for the `optionLabel`. Not a big deal, I noticed this while debugging the Storybook examples.

In the case of an empty string would it be better to fall back to the `label` value?
```suggestion
	const optionLabel = !! radioProps[ 'aria-label' ]
		? radioProps[ 'aria-label' ]
		: label;
```

File: packages/components/src/toggle-group-control/stories/index.js [hunk 7]
```diff
@@ -101,6 +106,16 @@ Default.args = {
...
 
+export const WithTooltipText = _default.bind( {} );
+WithTooltipText.args = {
+	...Default.args,
+	options: [
+		{ value: 1, label: '1', tooltipText: 'One' },
+		{ value: 2, label: '2', tooltipText: 'Two' },
+		{ value: 3, label: '3', tooltipText: 'Three' },
+	],
+};
```
Comment (aaronrobertshaw) [inline packages/components/src/toggle-group-control/stories/index.js]: This will need updating to use `aria-label` via `showTooltip` as well.

```suggestion
export const WithTooltip = _default.bind( {} );
WithTooltip.args = {
	...Default.args,
	options: [
		{ value: 1, label: '1', showTooltip: true, 'aria-label': 'One' },
		{ value: 2, label: '2', showTooltip: true, 'aria-label': 'Two' },
		{ value: 3, label: '3', showTooltip: true, 'aria-label': 'Three' },
	],
};
```

**Review by aaronrobertshaw — COMMENTED** [review]
This is looking pretty good so far @ramonjd, nice work ✨ 

✅ Unit tests pass
✅ Existing uses of the `ToggleGroupControl` in the editor still work
✅ Tooltips display when hoving with mouse or selecting via keyboard
❓ Storybook tooltip examples do not work (see comments below)

When switching to the `showTooltip` approach it appears as though the component's stories were missed.

File: packages/components/src/toggle-group-control/stories/index.js (hunk 5 above)
Comment (ramonjd) [reply packages/components/src/toggle-group-control/stories/index.js]: Thanks for catching that. Yes, I complete forgot to update the story!

**Comment by ramonjd** [comment]
Thanks for leaving the suggestions @aaronrobertshaw I managed to commit all this from my phone thanks to you ☎️

File: packages/components/src/toggle-group-control/toggle-group-control-option/component.tsx [hunk 8]
```diff
@@ -9,7 +9,6 @@ import { Radio } from 'reakit';
 /**
  * WordPress dependencies
  */
-import { __ } from '@wordpress/i18n';
```
Comment (ramonjd) [inline packages/components/src/toggle-group-control/toggle-group-control-option/component.tsx:12]: I couldn't see where this import was being used.

File: packages/components/CHANGELOG.md [hunk 9]
```diff
@@ -6,6 +6,9 @@
 
 -   Replaced hardcoded blue in `ColorPicker` with UI theme color ([#36153](https://github.com/WordPress/gutenberg/pull/36153)).
 
+### Enhancements
+-   Added a `showTooltip` prop to `ToggleGroupControlOption` in order to display tooltip text (using `<Tooltip />`). ([36726](https://github.com/WordPress/gutenberg/pull/36726)).
```
Comment (ciampo) [inline packages/components/CHANGELOG.md]: Tiniest on nits

```suggestion
-   Added a `showTooltip` prop to `ToggleGroupControlOption` in order to display tooltip text (using `<Tooltip />`). ([#36726](https://github.com/WordPress/gutenberg/pull/36726)).
```

File: packages/components/src/toggle-group-control/test/index.js [hunk 10]
```diff
@@ -38,4 +53,36 @@ describe( 'ToggleGroupControl', () => {
...
 		expect( mockOnChange ).toHaveBeenCalledWith( 'rigas' );
 	} );
+	it( 'should render tooltip where `showTooltip` === `true`', async () => {
+		render(
+			<ToggleGroupControl label="Test Toggle Group Control">
+				{ optionsWithTooltip }
+			</ToggleGroupControl>
+		);
+		const firstRadio = screen.getByLabelText(
+			'Click for Delicious Gnocchi'
```
Comment (ciampo) [inline packages/components/src/toggle-group-control/test/index.js:71]: Thank you for adding these unit tests!

(also, I wish the "click for gnocchi" was real!)

File: packages/components/src/tooltip/index.js (hunk 4 above)
Comment (ciampo) [reply packages/components/src/tooltip/index.js]: Got it, thank you for the insight!

I would suggest a slightly different approach which doesn't introduce any runtime changes, while still "tricking" the TS compiler not to display an error — this is achieved by moving the props destructuring inside the component's implementation:

```diff
diff --git a/packages/components/src/tooltip/index.js b/packages/components/src/tooltip/index.js
index deeb1c8c26..85859e1b17 100644
--- a/packages/components/src/tooltip/index.js
+++ b/packages/components/src/tooltip/index.js
@@ -92,13 +92,9 @@ const emitToChild = ( children, eventName, event ) => {
 	}
 };
 
-function Tooltip( {
-	children,
-	position,
-	text,
-	shortcut = null,
-	delay = TOOLTIP_DELAY,
-} ) {
+function Tooltip( props ) {
+	const { children, position, text, shortcut, delay = TOOLTIP_DELAY } = props;
+
 	/**
 	 * Whether a mouse is currently pressed, used in determining whether
 	 * to handle a focus event as displaying the tooltip immediately.
```

File: packages/components/src/toggle-group-control/test/index.js [hunk 11]
```diff
@@ -38,4 +53,36 @@ describe( 'ToggleGroupControl', () => {
...
+		const secondRadio = screen.getByLabelText(
+			'Click for Sumptuous Caponata'
+		);
+		fireEvent.mouseOver( secondRadio );
+		await waitFor( () =>
+			expect(
+				screen.queryByText( 'Click for Sumptuous Caponata' )
+			).not.toBeInTheDocument()
+		);
+	} );
```
Comment (ciampo) [inline packages/components/src/toggle-group-control/test/index.js]: Since the tooltip does not show immediately, "waiting for the tooltip text not to be in the document" will aways be a successful test (even when the tooltip is enabled).

Probably the easiest way could be to use fake timers ([RTL docs](https://testing-library.com/docs/using-fake-timers/), [Jest docs](https://jestjs.io/docs/next/timer-mocks))

Looks like you could do something like:

```js
it('test name', () => {
  jest.useFakeTimers();

  // do stuff, up until firing mouseOver event

  jest.runAllTimers();

  // Check for the tooltip text not to be in the document

  jest.runOnlyPendingTimers();
  jest.useRealTimers()
}
```

Although I've never used these APIs before, so I'm not sure if they will work as expected!

**Review by ciampo — COMMENTED** [review]
Thank you for the great work, Ramon! For some reason I must have been faster with my mouse than the 700ms tooltip delay 😅  I can confirm that it works as expected!

I've left a few more comments, but they're all minor, but we're very close!

File: packages/components/src/tooltip/index.js (hunk 4 above)
Comment (ntsekouras) [reply packages/components/src/tooltip/index.js]: >The TS was complaining that shortcut was a required prop, but, looking at all usages of Tooltip, it isn't.

Do we know why this was shown? I couldn't find anywhere in GB repo where this prop was deemed required. Same goes to `position` prop I believe.

File: packages/components/src/toggle-group-control/toggle-group-control-option/component.tsx (hunk 6 above)
Comment (ntsekouras) [reply packages/components/src/toggle-group-control/toggle-group-control-option/component.tsx]: No strong opinions here, but I had the same question when introduced and @ciampo suggested to have this check: https://github.com/WordPress/gutenberg/pull/35423#discussion_r724760813.

File: packages/components/src/tooltip/index.js (hunk 4 above)
Comment (ciampo) [reply packages/components/src/tooltip/index.js]: It's not deemed as required, but with the current notation TS _thinks_ it's required, since it doesn't have a default value (and since there isn't any alternative JSDoc to say otherwise).

That's why, if we move the prop destructuring to the function body, TS has to no way to infer the type of `props` and therefore can't throw any errors

Comment (ntsekouras) [reply packages/components/src/tooltip/index.js]: Thanks for explaining!

Comment (ramonjd) [reply packages/components/src/tooltip/index.js]: Thanks for the explanation and the destructuring example 🙇

File: packages/components/src/toggle-group-control/test/index.js (hunk 11 above)
Comment (ramonjd) [reply packages/components/src/toggle-group-control/test/index.js]: > Since the tooltip does not show immediately, "waiting for the tooltip text not to be in the document" will aways be a successful test (even when the tooltip is enabled).

Thanks @ciampo !

I should have looked at the [Tooltip tests](https://github.com/WordPress/gutenberg/blob/21c9f042c3c0f44ee6147809d74291b67724a395/packages/components/src/tooltip/test/index.js) for guidance too, where it uses the fake timers. It's much neater now!

File: packages/components/src/toggle-group-control/test/index.js (hunk 10 above)
Comment (ramonjd) [reply packages/components/src/toggle-group-control/test/index.js:71]: My Nonna is on speed dial. Same thing. 😆

File: packages/components/src/toggle-group-control/test/index.js [hunk 12]
```diff
@@ -7,24 +7,52 @@ import { render, fireEvent, screen } from '@testing-library/react';
...
 import { ToggleGroupControl, ToggleGroupControlOption } from '../index';
+import { TOOLTIP_DELAY } from '../../tooltip';
 
 describe( 'ToggleGroupControl', () => {
+	beforeEach( () => {
+		jest.useFakeTimers();
+	} );
+
+	afterEach( () => {
+		jest.useRealTimers();
```
Comment (ciampo) [inline packages/components/src/toggle-group-control/test/index.js]: The [docs from React Testing Library](https://testing-library.com/docs/using-fake-timers/) suggest to also add 

```suggestion
		jest.runOnlyPendingTimers();
		jest.useRealTimers();
```

**Review by ciampo — APPROVED** [review]
Thank you @ramonjd for addressing all the feedback!

I left one last minor comment, but feel free to merge once that's addressed too!

LGTM 🚀

**Review by diegohaz — APPROVED** [review]
Great work @ramonjd! Thank you!

I wonder if we should just show the tooltips without a delay since they're semantically labels, and not just tips.

Anyway, tooltips are hard! Here's an excellent video that explains several problems with tooltips: https://youtu.be/lb0_v7D4kbs

We may also consider a design where both the number and the text are visible without the need of tooltips.

But this PR is already an improvement over the current version, so let's merge it! 🚀

File: packages/components/src/toggle-group-control/test/index.js (hunk 12 above)
Comment (ramonjd) [reply packages/components/src/toggle-group-control/test/index.js]: Thanks for the suggestion.

The docs are ambiguous on this one I see. See https://jestjs.io/docs/timer-mocks#run-pending-timers, which loosely recommends running `jest.runOnlyPendingTimers();` to clear regular intervals.  

I understand that we'd want to flush out pending timers though, and I did try using `runOnlyPendingTimers` originally but it caused the tests to fail, and I'm not really sure why. 

This was also after trying to remove `setTimeout` from the test itself and use `jest.advanceTimersByTime( TOOLTIP_DELAY );` and `runOnlyPendingTimers` within the test. `jest.clearAllTimers()` in the `afterEach` worked as a fallback.

`<Tooltip />` uses `useDebounce` so it might be due to me not testing correctly.

🎉 🎉 🎉 🎉 🎉 🎉 🎉 

Anyway, there's a happy ending. It turns out that `<Tooltip />` only debounces `onMouseEnter` and not `focus` so I switched to testing using that event. 

It's cleaner, and with no `setTimeout`, no fake timers, and `<Tooltip />` already has test coverage for `mouseEnter|Leave`.  

WOOT!

**Comment by ramonjd** [comment]
Thanks for the ✅ @ciampo and for the info and the helpful link @diegohaz 

> I wonder if we should just show the tooltips without a delay since they're semantically labels, and not just tips.

It looks like there's only a delay when we mouse over a tooltip. See: https://github.com/WordPress/gutenberg/blob/4ef7fc547bbc0eef8e445b45afee46f5ca363c18/packages/components/src/tooltip/index.js#L195, where `true` tells the event handler to debounce the tooltip.

Seems like a pretty easy change to make assuming there's consensus on removing it. Maybe the delay was there to reduce any perceived "clutter" for power speed mouse users like @ciampo and me :trollface: 

Once the tests pass I'll 🚢

File: packages/components/CHANGELOG.md [hunk 13]
```diff
@@ -6,19 +6,22 @@
...
+
 ### Experimental
 
--   Reinstated the ability to pass additional props to the `ToolsPanel` ([36428](https://github.com/WordPress/gutenberg/pull/36428)).
+-   Reinstated the ability to pass additional props to the `ToolsPanel` ([#36428](https://github.com/WordPress/gutenberg/pull/36428)).
 -   Added an `__unstable-large` size variant to `InputControl`, `SelectControl`, and `UnitControl` for selective migration to the larger 40px heights. ([#35646](https://github.com/WordPress/gutenberg/pull/35646)).
 -   Fixed inconsistent padding in `UnitControl` ([#35646](https://github.com/WordPress/gutenberg/pull/35646)).
 -   Added support for RTL behavior for the `ZStack`'s `offset` prop ([#36769](https://github.com/WordPress/gutenberg/pull/36769))
--   Fixed race conditions causing conditionally displayed `ToolsPanelItem` components to be erroneously deregistered ([36588](https://github.com/WordPress/gutenberg/pull/36588)).
+-   Fixed race conditions causing conditionally displayed `ToolsPanelItem` components to be erroneously deregistered ([#36588](https://github.com/WordPress/gutenberg/pull/36588)).
```
Comment (ramonjd) [inline packages/components/CHANGELOG.md:18]: Just tidying up the docs and adding a `#` in the link text, which represent PRs

File: packages/components/src/toggle-group-control/test/index.js (hunk 12 above)
Comment (ciampo) [reply packages/components/src/toggle-group-control/test/index.js]: That's great, thank you!
